File: docs/templates/stacks/python/security.py
from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse


def install_security(app: FastAPI, service_name: str) -> None:
    @app.middleware("http")
    async def security_middleware(request: Request, call_next):
        if request.url.path.startswith("/health/"):
            return await call_next(request)

        auth = authenticate_request(request)
        if auth is None:
            return JSONResponse(status_code=401, content={"code": "unauthorized", "message": "Bearer token is invalid or missing."})
        if request.method not in {"GET", "HEAD", "OPTIONS"} and not request.headers.get("x-correlation-id"):
            return JSONResponse(status_code=400, content={"code": "correlation_id_required", "message": "Mutation requests require X-Correlation-Id."})
        if not authorize_request(service_name, request, auth):
            return JSONResponse(status_code=403, content={"code": "forbidden", "message": "Request is not authorized."})

        request.scope["erp_auth"] = auth
        return await call_next(request)


def authenticate_request(request: Request) -> dict | None:
    return None


def authorize_request(service_name: str, request: Request, auth: dict) -> bool:
    return bool(service_name and auth.get("subject"))

File: docs/templates/stacks/typescript/security.ts
import { IncomingMessage, ServerResponse } from "node:http";

export type AuthContext = {
  subject: string;
  tenantSlug: string;
  scopes: string[];
};

export async function enforceSecurity(
  serviceName: string,
  request: IncomingMessage,
  response: ServerResponse,
  next: () => Promise<void> | void
): Promise<void> {
  if (request.url?.startsWith("/health/")) {
    await next();
    return;
  }

  const auth = authenticateRequest(request);
  if (!auth) {
    writeError(response, 401, "unauthorized", "Bearer token is invalid or missing.");
    return;
  }
  if (!["GET", "HEAD", "OPTIONS"].includes(request.method ?? "GET") && !request.headers["x-correlation-id"]) {
    writeError(response, 400, "correlation_id_required", "Mutation requests require X-Correlation-Id.");
    return;
  }
  if (!authorizeRequest(serviceName, request, auth)) {
    writeError(response, 403, "forbidden", "Request is not authorized.");
    return;
  }

  request.headers["x-erp-auth-subject"] = auth.subject;
  request.headers["x-erp-auth-tenant"] = auth.tenantSlug;
  request.headers["x-erp-auth-scopes"] = auth.scopes.join(" ");
  await next();
}

function authenticateRequest(_request: IncomingMessage): AuthContext | null {
  return null;
}

function authorizeRequest(serviceName: string, _request: IncomingMessage, auth: AuthContext): boolean {
  return serviceName.length > 0 && auth.subject.length > 0;
}

function writeError(response: ServerResponse, status: number, code: string, message: string): void {
  response.writeHead(status, { "content-type": "application/json" });
  response.end(JSON.stringify({ code, message }));
}
